database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -26,69 +28,72 @@
  * This class implements an array-like collection on top of a Firebase location.
  */
 class FirebaseArray implements ChildEventListener {
-    public interface OnChangedListener {
+    public interface ChangeEventListener {
         enum EventType {ADDED, CHANGED, REMOVED, MOVED}
 
         void onChanged(EventType type, int index, int oldIndex);
 
-        void onCancelled(DatabaseError databaseError);
+        void onCancelled(DatabaseError error);
     }
 
+    protected ChangeEventListener mListener;
+    protected boolean mIsListening;
     private Query mQuery;
-    private OnChangedListener mListener;
     private List<DataSnapshot> mSnapshots = new ArrayList<>();
 
     public FirebaseArray(Query ref) {
         mQuery = ref;
+    }
+
+    public void setChangeEventListener(@NonNull ChangeEventListener listener) {
+        if (mIsListening && listener == null) {
+            throw new IllegalStateException("Listener cannot be null.");
+        }
+        mListener = listener;
+    }
+
+    public void startListening() {
+        if (mListener == null) throw new IllegalStateException("Listener cannot be null.");
         mQuery.addChildEventListener(this);
+        mIsListening = true;
     }
 
-    public void cleanup() {
+    public void stopListening() {
         mQuery.removeEventListener(this);
+        mSnapshots.clear();
+        mIsListening = false;
     }
 
-    public int getCount() {
+    public int size() {
```

Comment:
These look like pretty breaking changes, but I'm only just someone really curious about this project.
Also, was a consensus ever reached on using enums in the code? Shouldn't we be doing something like this:

    public interface ChangeListener {
        public static class EventType {
            public static final int ADDED = 1;
            public static final int CHANGED = 2;
            // ...
        }
    }